plugin/plugin.go

```diff
@@ -0,0 +1,188 @@
+	CreateChangeEvent bool   `envconfig:"PLUGIN_CREATE_CHANGE_EVENT"`
+	Resolve           bool   `envconfig:"PLUGIN_RESOLVE"`
+	JobStatus         string `envconfig:"PLUGIN_JOB_STATUS"`
+	CustomDetailsStr  string `envconfig:"PLUGIN_CUSTOM_DETAILS"` // Intermediate string to receive JSON
+	CustomDetails     map[string]interface{}
+}
+
+// PagerDutyClient defines the methods used from the PagerDuty API.
+type PagerDutyClient interface {
+	ManageEventWithContext(ctx context.Context, event *pagerduty.V2Event) (*pagerduty.V2EventResponse, error)
+	CreateChangeEventWithContext(ctx context.Context, event pagerduty.ChangeEvent) (*pagerduty.ChangeEventResponse, error)
+}
+
+// Exec executes the plugin.
+func Exec(ctx context.Context, args Args) error {
+	client := pagerduty.NewClient(args.RoutingKey)
+	logger := logrus.
+		WithField("PLUGIN_ROUTING_KEY", args.RoutingKey).
+		WithField("PLUGIN_INCIDENT_SUMMARY", args.IncidentSummary).
+		WithField("PLUGIN_INCIDENT_SOURCE", args.IncidentSource).
+		WithField("PLUGIN_INCIDENT_SEVERITY", args.IncidentSeverity).
+		WithField("PLUGIN_CREATE_CHANGE_EVENT", args.CreateChangeEvent).
+		WithField("PLUGIN_JOB_STATUS", args.JobStatus)
+
+	logger.Info("Starting plugin execution")
+
+	// Validate required fields
+	if args.RoutingKey == "" {
+		return errors.New("routingKey is required")
+	}
+	if args.IncidentSummary == "" {
+		return errors.New("incidentSummary is required")
+	}
+	if args.IncidentSource == "" {
+		return errors.New("incidentSource is required")
+	}
+	if !args.CreateChangeEvent {
+		if args.DedupKey == "" {
+			return errors.New("dedupKey is required when not creating a change event")
+		}
+		if args.JobStatus == "" {
+			return errors.New("jobStatus is required when not creating a change event")
+		}
+	}
+
+	if args.JobStatus == "" {
+		logger.Warn("Job status is empty")
+	}
+
+	if args.CreateChangeEvent {
+		logger.Info("Creating change event")
+		if err := createChangeEvent(ctx, client, args); err != nil {
+			logger.WithError(err).Error("Failed to create change event")
+			return errors.New("failed to create change event")
+		}
+		logger.Info("Change event created Successfully")
+		return nil
+	}
+
+	// Handle job status and decide whether to trigger or resolve incidents
+	var resolveIncident bool
+	var severity = args.IncidentSeverity
+	var summary = args.IncidentSummary
+
+	switch args.JobStatus {
+	case "success":
+		resolveIncident = args.Resolve
+		summary = "Job succeeded: " + summary
+		logger.Info("Job succeeded, deciding on resolving incident")
+	case "failure":
+		resolveIncident = false
+		severity = "critical"
+		summary = "Job failed: " + summary
+		logger.Info("Job failed, deciding on triggering or resolving incident")
+	case "unstable":
+		resolveIncident = false
+		severity = "warning"
+		summary = "Job is unstable: " + summary
+		logger.Info("Job is unstable, deciding on triggering or resolving incident")
+	case "aborted":
+		resolveIncident = false
+		severity = "warning"
```

Comment:
define constants and call those as needed:
```
const (
    SeverityCritical = "critical"
    SeverityWarning  = "warning"
    SeverityInfo     = "info"
)
```

also, are these hardcoded severities mappings depending on jobStatus predefined somewhere in PagerDuty? just want to ensure plugin complies with PagerDuty job status and severity levels.

please clarify in the code comments about these severities mappings.

Reply:
Note: The severity mappings ("success", "failure", etc.) are not predefined by PagerDuty.
As per this [doc](https://response.pagerduty.com/before/severity_levels/), severity levels depend on the severity of the issue.
I have removed those hardcoded severity levels, It is better to allow users to pass the severity value as an environment variable to ensure compliance with PagerDuty's guidelines. The severity can be taken from the jenkins trace from the param "incidentSeverity" in conversion.